### ISSUE TITLE
Enum type is lost when used in attribute

### ISSUE THREAD

**Issue by Novakov** [dotnet/fsharp#995]
**Description**
Enum type is lost (casted to int32) when used in attribute named argument of type `obj`.

http://stackoverflow.com/questions/35732830/how-to-preserve-value-type-in-enum

**Repro Steps**
Create attribute with named argument of type `obj`. Add that attribute and assign enum value to named argument.

``` fsharp
type MyAttribute() =
    inherit Attribute()

    let mutable prop = null

    member this.Prop
        with get(): obj = prop
        and  set(value) = prop <- value


type MyEnum = 
    | A = 1
    | B = 2

[<My(Prop = MyEnum.B)>]
type MyClass = class
    end

let test () =
   let t = typeof<MyClass>
   let a = t.GetCustomAttributes(false).[0] :?> MyAttribute

   let e = a.Prop
   Convert.ToString(e, Globalization.CultureInfo.CurrentCulture)
```

**Expected**
`test` function should return `"B"`. In IL value should be stored as enum.

**Actual**
`test` function returns `2`. In IL value is stored as int32 - enum type is lost

**Version**: `Microsoft (R) F# Compiler version 14.0.23413.0`

**Comment by gsomix** [comment]
Hey, any news on it? Thanks!

**Comment by abelbraaksma** [comment]
Why is this actually marked as feature request? Is it by design that the type information is lost, or is it because the underlying type of the enum is int?

**Comment by isaevdan** [comment]
As for me this looks like a bug, and this does not happen in similar C# code

**Comment by cartermp** [comment]
@dsyme to comment on the labeling, I personally think it's a bug

**Comment by dsyme** [comment]
Yes, it can be considered a bug, though it's fix might need an RFC

The translation from Enums to constants is done too early. IIRC the only reason I didn't fix it was a concern that quotations actually see the constant, not the field, and because it would mean storing more information in the TAST which we have to do very carefully w.rt. compat.

We can fix this, but it may need a new quotation node that refines "Const" - that is, the node will continue to match "BasicPatterns.Const" but there is an additional "BasicPatterns.EnumConst" or something that sees this refined information.
